File: go/backup.go
package main

import (
	"fmt"
	"log"
	"os"
	"github.com/algolia/algoliasearch-client-go/v4/algolia/search"
	"github.com/joho/godotenv"
	"encoding/json"
)

func PrintErrAndExit(err error) {
	fmt.Println(err)
	os.Exit(1)
}


func PrintCurrentObjects() {
	appID, apiKey, indexName := os.Getenv("ALGOLIA_APP_ID"), os.Getenv("ALGOLIA_API_KEY"), os.Getenv("ALGOLIA_INDEX_NAME")
	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		// The client can fail to initialize if you pass an invalid parameter.
		fmt.Printf("Client error")
		panic(err)
	}
	resSearch, err := client.SearchSingleIndex(client.NewApiSearchSingleIndexRequest(indexName))
	if err != nil {
		// handle the eventual error
		panic(err)
	}
	fmt.Println("Current objects: ", resSearch.Hits, "\n")
}

func main() {
	err := godotenv.Load()
	if err != nil {
		log.Fatal("Error loading .env file")
	}
	// Algolia client credentials
	appID, apiKey, indexName := os.Getenv("ALGOLIA_APP_ID"), os.Getenv("ALGOLIA_API_KEY"), os.Getenv("ALGOLIA_INDEX_NAME")

	// Start the API client
	// https://www.algolia.com/doc/libraries/sdk/methods/search#go
	client, err := search.NewClient(appID, apiKey)
	if err != nil {
		// The client can fail to initialize if you pass an invalid parameter.
		fmt.Printf("Client error")
		panic(err)
	}

	// Get all records from an index
	// https://www.algolia.com/doc/api-reference/api-methods/browse/#get-all-records-from-an-index
	// Use an API key with `browse` ACL
	resBrowseObj, err := client.Browse(client.NewApiBrowseRequest(
		indexName))

	if err != nil {
	// handle the eventual error
		panic(err)
	}

	fmt.Println("Current objects")
	PrintCurrentObjects()

	// Encode array to json
	resEncodeObjJson, err := json.Marshal(resBrowseObj.Hits)
	if err != nil {
		log.Fatal(err)
	}

	// Write json to file
	err = os.WriteFile(indexName + "_records.json", resEncodeObjJson, 0644) // 0644 gives read/write permissions for the owner, and read for others
	if err != nil {
		log.Fatalf("Error writing to file: %v", err)
	}

	// Retrieve settings for an index
	// https://www.algolia.com/doc/api-reference/api-methods/get-settings/#retrieve-settings-for-an-index
	resGetSettings, err := client.GetSettings(client.NewApiGetSettingsRequest(
	indexName).WithGetVersion(2))
		if err != nil {
		// handle the eventual error
		panic(err)
	}

	fmt.Println("Current settings")
	fmt.Println(resGetSettings)

	// Encode array to json
	resEncodeSettingsJson, err := json.Marshal(resGetSettings)
	if err != nil {
		log.Fatal(err)
	}
	
	// Write json to file
	err = os.WriteFile(indexName + "_settings.json", resEncodeSettingsJson, 0644) // 0644 grants read/write permissions for the owner
	if err != nil {
		log.Fatalf("Error writing to file: %v", err)
	}

	// Export rules 
	// https://www.algolia.com/doc/api-reference/api-methods/export-rules/
	var rules []search.Rule

	err = client.BrowseRules(
		indexName,
		*search.NewEmptySearchRulesParams(),
		search.WithAggregator(func(r any, err error) {
			if err != nil {
				log.Fatalf("There was an error: %v", err)
			}

			rules = append(rules, r.(*search.SearchRulesResponse).Hits...)
		}),
	)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Current rules:")
	for _, rule := range rules {
		fmt.Printf("- Rule: %s\n", rule.ObjectID)
	}

	// Encode array to json
	resEncodeRulesJson, err := json.Marshal(rules)
	if err != nil {
		log.Fatal(err)
	}

	// Write json to file
	err = os.WriteFile(indexName + "_rules.json", resEncodeRulesJson, 0644) // 0644 grants read/write permissions for the owner
	if err != nil {
		log.Fatalf("Error writing to file: %v", err)
	}

	// Export synonyms
	// https://www.algolia.com/doc/api-reference/api-methods/export-synonyms/
		var synonyms []search.SynonymHit

	err = client.BrowseSynonyms(
		indexName,
		*search.NewEmptySearchSynonymsParams(),
		search.WithAggregator(func(r any, err error) {
			if err != nil {
				log.Fatalf("There was an error: %v", err)
			}

			synonyms = append(synonyms, r.(*search.SearchSynonymsResponse).Hits...)
		}),
	)
	if err != nil {
		log.Fatal(err)
	}

	for _, synonym := range synonyms {
		fmt.Printf("- Synonym: %s\n", synonym.ObjectID)
	}

	// Encode array to json
	resEncodeSynonymsJson, err := json.Marshal(synonyms)
	if err != nil {
		log.Fatal(err)
	}

	// Write json to file
	err = os.WriteFile(indexName + "_synonyms.json", resEncodeSynonymsJson, 0644) // 0644 grants read/write permissions for the owner
	if err != nil {
		log.Fatalf("Error writing to file: %v", err)
	}
}

File: java/src/main/java/Backup.java
import java.io.File;
import java.io.IOException;
import java.util.concurrent.ExecutionException;

import com.algolia.api.SearchClient;
import com.algolia.model.search.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.node.ObjectNode;

import io.github.cdimascio.dotenv.Dotenv;

public class Backup {
    
    public static void main(String[] args) throws ExecutionException, InterruptedException, IOException {
        Backup.run();
    }

    public static void run() throws ExecutionException, InterruptedException, IOException {
    
        Dotenv dotenv = Dotenv.configure().load();

        String appID = dotenv.get("ALGOLIA_APP_ID");
        String apiKey = dotenv.get("ALGOLIA_API_KEY");
        String indexName = dotenv.get("ALGOLIA_INDEX_NAME");

        // Start the API client
        // https://www.algolia.com/doc/libraries/sdk/methods/search#java
        try (SearchClient client = new SearchClient(appID, apiKey)) {

            // Initialize ObjectMapper
            ObjectMapper objectMapper = new ObjectMapper();
            objectMapper.enable(SerializationFeature.INDENT_OUTPUT);

            // Get all records from an index
            // https://www.algolia.com/doc/api-reference/api-methods/browse/#get-all-records-from-an-index
            // Use an API key with `browse` ACL
            BrowseResponse<Hit> respRecords = client.browse(indexName, Hit.class);

            // Write records to file
            File recordsFile = new File(indexName + "_records.json");
            objectMapper.writeValue(recordsFile, respRecords.getHits());      

            // Retrieve settings for an index
            // https://www.algolia.com/doc/api-reference/api-methods/get-settings/#retrieve-settings-for-an-index

            // Export settings
            SettingsResponse respSettings = client.getSettings(indexName, 2);

            // Write settings to file
            File settingsFile = new File(indexName + "_settings.json");
            objectMapper.writeValue(settingsFile, respSettings);

            // The following settings will export as null if the default value is used and the setting hasn't been modified
            // typoTolerance, ignorePlurals, removeStopWords, distinct, reRankingApplyFilter, primary
            // Remove these settings if the values are null to prevent errors while restoring

            //Get the file
            ObjectNode settingsFileContent = (ObjectNode) objectMapper.readTree(settingsFile);

            // Create array of settings
            String[] settingsValues = {"typoTolerance", "ignorePlurals", "removeStopWords", "distinct", "reRankingApplyFilter", "primary"};

            for (int i = 0; i < settingsValues.length; i++) {
                if (settingsFileContent.has(settingsValues[i]) && settingsFileContent.get(settingsValues[i]).isNull()) {
                    settingsFileContent.remove(settingsValues[i]);
                }
            }

            objectMapper.writeValue(settingsFile, settingsFileContent);

            // Export rules 
            // https://www.algolia.com/doc/api-reference/api-methods/export-rules/
            Iterable<Rule> respRules = client.browseRules(indexName, new SearchRulesParams());

            // Transform rules iterator to Array and then to JSON
            File rulesFile = new File(indexName + "_rules.json");
            objectMapper.writeValue(rulesFile, respRules);        

            // Export synonyms
            // https://www.algolia.com/doc/api-reference/api-methods/export-synonyms/
            Iterable<SynonymHit> respSynonyms = client.browseSynonyms(indexName, new SearchSynonymsParams());

            // Transform synonyms iterator to Array and then to JSON
            File synonymsFile = new File(indexName + "_synonyms.json");
            objectMapper.writeValue(synonymsFile, respSynonyms);  

        }
    }
}
